### PR TITLE
CM_Mail: Debug appending key/value to array

### PR DESCRIPTION
**Pull request by ppp0**

### CODE REVIEW

**Comment by ppp0** [comment]
@njam please review

**Comment by tomaszdurka** [comment]
Looks ok, but...

You're dropping ability to pass multiple headers with the same. According to http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2 it is possible to send headers with the same name as single with values separated by commas. Should we support this? I vote for no.

The only misleading thing now name of the method. We should probably rename it to "setCustomHeader" as it will actually replace previoous one (not add).

**Comment by njam** [comment]
Agree, I would keep the ability to set multiple headers with the same key.

**Comment by ppp0** [comment]
I gather from this:
- @tomaszdurka you are ok with the current solution (unique header name) but want a renaming of the setter
- @njam you agree but would like to keep setting multiple headers with the same name

this is an interest conflict..

I vote for multiple headers with the same name, which is standard.. in order to do that, I'd refactor to encaspulate every single k/v pair into a single array (and therefore have an array of arrays). I need to refactor the "explosion" into single headers - easy :v:

**Comment by ppp0** [comment]
@tomaszdurka @njam who wants to review?

**Comment by njam** [comment]
Philipp and me are looking into this now

**Comment by tomaszdurka** [comment]
I would prefer to not keep array-like types with 2 values. Rather:
```php
if (!isset($this->_customHeaders[$label])) {
  $this->_customHeaders[$label] = array();
}
$this->_customHeaders[$label][] = $value;
```

**Comment by tomaszdurka** [comment]
@ppp0 I just meant that if we agree on single-header then the name sould be set* as it's replaces previous value.

**Comment by ppp0** [comment]
@tomaszdurka @njam I value your comment aka please review

File: tests/library/CM/MailTest.php [hunk 1]
```diff
@@ -74,28 +77,64 @@ public function testSetText() {
         $this->assertSame($text, $mail->getText());
     }
 
+    public function getMockPhpMailer() {
+        return $this->_mockPHPMailer;
```
Comment (mariansollmann) [inline tests/library/CM/MailTest.php]: If you have a private variable why you need this getter?

File: tests/library/CM/MailTest.php [hunk 2]
```diff
@@ -74,28 +77,64 @@ public function testSetText() {
...
-            $this->fail('Sent mail without recipient');
-        } catch (Exception $e) {
-            $this->assertContains('No recipient specified', $e->getMessage());
-        }
-        $mail->addTo('tomasz@durka.pl');
-        $mail->setSubject('foo');
-        $mail->setText('bar');
-        $this->assertSame(0, CM_Mail::getQueueSize());
+        CM_Config::get()->CM_Mail->send = true;
+        $this->_mockPHPMailer = $this->mockObject('PHPMailer');
```
Comment (mariansollmann) [inline tests/library/CM/MailTest.php]: I think this should go to the `setUp`, wdyt?

File: tests/library/CM/MailTest.php (hunk 1 above)
Comment (ppp0) [inline tests/library/CM/MailTest.php]: I call this method from another class `Test_CM_Mail` (line 190++) - for this I need the getter?

File: tests/library/CM/MailTest.php [hunk 3]
```diff
@@ -74,28 +77,64 @@ public function testSetText() {
...
+                $this->assertSame('bar,foo', $value);
+            })
+            ->at(1, function ($name, $value) {
+                $this->assertSame('X-Bar', $name);
+                $this->assertSame('foo', $value);
+            });
+
+        $sendMethod = $phpMailer->mockMethod('Send');
+
+        $mail = new Test_CM_Mail('foo@example.com', null, null, $this);
```
Comment (mariansollmann) [inline tests/library/CM/MailTest.php]: Why you are using a new class? Couldn't you just mock `CM_Mail` and "overwrite" `_getPHPMailer` with your mock `_mockPHPMailer`, wdyt?

**Comment by ppp0** [comment]
@mariansollmann fancy reviewing please?

**Comment by mariansollmann** [comment]
looks good

**Comment by njam** [comment]
Why is it now called `setCustomHeader()`, isn't it *adding* a header (i.e. we can have twice a header with the same name)?

**Comment by ppp0** [comment]
It is overwriting the header's key, though

**Comment by njam** [comment]
...but only if it's the *same*, so not really overwriting :)

**Pull request merged**